### PR TITLE
Add inference latency test tool

### PR DESCRIPTION
**Pull request by RunningLeon**
## Motivation

Test model latency with a single image with fixed shape for a number of iterations.

## Modification

- [x] Add `tools/profile.py`
- [x] Update document
- [x] Update test

## BC-breaking (Optional)

None

####  Tensorrt model
```shell
python tools/profile.py \
    configs/mmcls/classification_tensorrt_dynamic-224x224-224x224.py \
    ../mmclassification/configs/resnet/resnet18_8xb32_in1k.py \
    ../mmdetection/demo
    --model work-dirs/mmcls/resnet/trt/end2end.engine \
    --device cuda \
    --shape 224x224 \
    --num-iter 100 \
    --warmup 10 \
```

#### PyTorch model
```
python tools/profile.py \
    configs/mmseg/segmentation_tensorrt-int8_static-512x512.py \
    ../mmsegmentation/configs/pspnet/pspnet_r50-d8_512x1024_40k_cityscapes.py \
    ../mmsegmentation/demo \
    --model [work-dirs/mmcls/resnet/trt/end2end.engine](https://download.openmmlab.com/mmsegmentation/v0.5/pspnet/pspnet_r50-d8_512x1024_40k_cityscapes/pspnet_r50-d8_512x1024_40k_cityscapes_20200605_003338-2966598c.pth) \
    --device cuda \
    --shape 512x512 \
    --num-iter 100 \
    --warmup 10 \
```
#### Output message
```text
----- Settings:
+------------+---------+
| batch size |    1    |
|   shape    | 224x224 |
| iterations |   100   |
|   warmup   |    10   |
+------------+---------+
----- Results:
+--------+------------+---------+
| Stats  | Latency/ms |   FPS   |
+--------+------------+---------+
|  Mean  |   1.535    | 651.656 |
| Median |   1.665    | 600.569 |
|  Min   |   1.308    | 764.341 |
|  Max   |   1.689    | 591.983 |
+--------+------------+---------+
```

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by irexyc** [comment]
For backend model, I think the api is bit complicated.

Since we only count the model execute time,  is it better to remove the deploy_cfg/model_cfg and add a backend option?

**Comment by AllentDan** [comment]
Shall we test Pytorch and Backend model both and give a comparison of them?

**Comment by AllentDan** [comment]
Shall we give deprecate warnings to `deploy.py` as speed test functions of `deploy.py` and `profile.py` are duplicated?

**Comment by RunningLeon** [comment]
@irexyc @AllentDan Hi, Any other comment on this PR?

File: tools/profile.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+def parse_args():
+    parser = argparse.ArgumentParser(
+        description='MMDeploy Model Latency Test Tool.')
+    parser.add_argument('deploy_cfg', help='Deploy config path')
+    parser.add_argument('model_cfg', help='Model config path')
+    parser.add_argument('image_dir', help='Input directory to image files')
+    parser.add_argument(
+        '--model', type=str, nargs='+', help='Input model files.')
+    parser.add_argument(
+        '--device', help='device used for conversion', default='cuda:0')
```
Comment (AllentDan) [inline tools/profile.py:27]: not conversion here

File: tools/profile.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        'be overwritten is a list, it should be like key="[a,b]" or key=a,b '
+        'It also allows nested list/tuple values, e.g. key="[(a,b),(c,d)]" '
+        'Note that the quotation marks are necessary and that no white space '
+        'is allowed.')
+
+    args = parser.parse_args()
+    return args
+
+
+def get_images(image_dir, extensions=['.jpg', '.jpeg', '.png']):
```
Comment (AllentDan) [inline tools/profile.py:58]: other image form like `bmp`?

File: mmdeploy/utils/timer.py [hunk 3]
```diff
@@ -127,3 +129,32 @@ def activate(cls,
...
+        results = PrettyTable()
+        results.field_names = ['Stats', 'Latency/ms', 'FPS']
+        results.add_rows([
+            ['Mean', latency_mean, fps_mean],
+            ['Median', latency_median, fps_median],
+            ['Min', latency_min, fps_min],
+            ['Max', latency_max, fps_max],
+        ])
+        results.float_format = '.3'
+        print(results)
```
Comment (AllentDan) [inline mmdeploy/utils/timer.py:160]: Logging methods of mmcv and mmdet prefer not using `print`

Comment (RunningLeon) [reply mmdeploy/utils/timer.py:160]: print is ok with me. Same as mmdet: https://github.com/open-mmlab/mmdetection/blob/56e42e72cdf516bebb676e586f408b98f854d84c/tools/test.py#L268

**Review by AllentDan — APPROVED** [review]

**Review by irexyc — APPROVED** [review]
Warmup and log_interval seems too small and the fps in log is unstable.

**Comment by RunningLeon** [comment]
> Warmup and log_interval seems too small and the fps in log is unstable.

default value may not be optimal for all tasks. User should change the setting accordingly.

**Comment by abdulazizab2** [comment]
I am getting errors when passing similar configs for testing trt model.
`python tools/profile.py "$MMDEPLOY_DIR"/configs/mmdet/detection/detection_tensorrt-fp16_dynamic-320x320-1344x1344.py "$MMDET_DIR"/configs/faster_rcnn/faster_rcnn_r50_fpn_2x_coco.py ${IMAGES_DIR} --model "$MMDEPLOY_DIR"/faster_rcnn_r50_fpn_2x_coco_tensorrt-fp16_dynamic-320x320-1344x1344/end2end.engine  --device cuda --shape 756x1344 --num-iter 100 --warmup 10`
I get the following error:
```
Traceback (most recent call last):
  File "tools/profile.py", line 70, in <module>
    class TorchWrapper(torch.nn.Module):
  File "tools/profile.py", line 78, in TorchWrapper
    def forward(self, *args, **kwargs):
  File "/home/abdulaziz/miniconda3/envs/mmdeploy/lib/python3.8/site-packages/mmdeploy/utils/timer.py", line 32, in _register
    assert warmup >= 1
TypeError: '>=' not supported between instances of 'str' and 'int'
```
I tried to debug and inspect. The warmup value is a string 'pytorch'.

**Comment by RunningLeon** [comment]
@abdulazizab2 Hi, could you post here which commit id of mmdeploy you're using? Cannot reproduce your problem on the latest master branch.

**Comment by abdulazizab2** [comment]
@RunningLeon 
`31cce6d16414b1452e5ae7092fde2450ecee727a`
I also check the difference between HEAD:tools/profile to origin:tools/profile and there are no changes FYI.

**Comment by RunningLeon** [comment]
@abdulazizab2 Hi, maybe you are using prebuilt old mmdeploy and because `name` argument is added to [here](https://github.com/open-mmlab/mmdeploy/blob/83b11bc1ca7227497928a57b56653b76501b1368/mmdeploy/utils/timer.py#L20). Could you use latest master branch and build from source?

**Comment by abdulazizab2** [comment]
@RunningLeon 
It works now. I had two different environments building mmdeploy. One was built with an older mmdeploy I suppose. Thank you!

**Comment by RunningLeon** [comment]
@abdulazizab2 Glad to know it solved. BR.

**Pull request merged**